File: spec/factories/locality.rb
# frozen_string_literal: true

require 'faker'

FactoryBot.define do
  factory :locality do
    school_name do
      Faker::Educator.university if user&.role_teacher?
    end
    school_type do
      if user&.role_teacher?
        school_types = Locality.school_types.keys
        school_types[SecureRandom.random_number(school_types.length - 1)]
      end
    end
    school_type_custom do
      'Werkrealschule' if school_type.try(:to_sym) == :school_type_custom
    end

    postal_code do
      postcode = Faker::Address.postcode
      country.in?(%i[AUT CHE]) ? postcode[0..3] : postcode
    end
    city { Faker::Address.city }
    country do
      countries = CountryService.codes
      countries[SecureRandom.random_number(countries.length - 1)]
    end

    state do
      states = CountryService.find(country).states.keys
      states[rand(SecureRandom.random_number(states.length - 1))] if states.any?
    end
  end
end

File: frontend/models/locality.js
import PropTypes from "prop-types";
import { attr, fk } from "redux-orm";

import BaseModel from ".";

class Locality extends BaseModel {}
Locality.modelName = "Locality";
Locality.collectionKey = "localities";

Locality.fields = {
  ...BaseModel.fields,
  schoolType: attr(),
  schoolTypeCustom: attr(),
  schoolName: attr(),
  country: attr(),
  state: attr(),
  city: attr(),
  postalCode: attr(),
  user: fk("User"),
};

Locality.defaultProps = {
  ...BaseModel.defaultProps,
  schoolType: null,
  schoolTypeCustom: null,
  schoolName: null,
  country: null,
  state: null,
  city: null,
  postalCode: null,
};

const Shape = PropTypes.shape({
  schoolType: PropTypes.string,
  schoolTypeCustom: PropTypes.string,
  schoolName: PropTypes.string,
  country: PropTypes.string,
  state: PropTypes.string,
  city: PropTypes.string,
  postalCode: PropTypes.string,
});

export default Locality;
export { Shape };
